test/unit_tests/api/test_processes.py

```diff
@@ -593,3 +595,116 @@ def unauthorized_workflow():
+
+    @inputstep("authorized_resume", assignee=Assignee.SYSTEM, resume_auth_callback=allow)
+    def authorized_resume(state):
+        user_input = yield ConfirmForm
+        return user_input.model_dump()
+
+    @inputstep("noauth_resume", assignee=Assignee.SYSTEM)
+    def noauth_resume(state):
+        user_input = yield ConfirmForm
+        return user_input.model_dump()
+
+    @workflow("test_auth_workflow", target=Target.CREATE)
+    def test_auth_workflow():
+        return init >> noauth_resume >> authorized_resume >> unauthorized_resume >> done
+
+    with WorkflowInstanceForTests(test_auth_workflow, "test_auth_workflow"):
+        response = test_client.post("/api/processes/test_auth_workflow", json=[{}])
+        assert HTTPStatus.CREATED == response.status_code
+        process_id = response.json()["id"]
+        # No auth succeeds
+        response = test_client.put(f"/api/processes/{process_id}/resume", json=[{"confirm": True}])
+        assert HTTPStatus.NO_CONTENT == response.status_code
+        # Authorized succeeds
+        response = test_client.put(f"/api/processes/{process_id}/resume", json=[{"confirm": True}])
+        assert HTTPStatus.NO_CONTENT == response.status_code
+        # Unauthorized fails
+        response = test_client.put(f"/api/processes/{process_id}/resume", json=[{"confirm": True}])
+        assert HTTPStatus.FORBIDDEN == response.status_code
+
+    # TODO test how this interacts with passing a different callback to @authorize_workflow
+    # These should be as functionally independent as possible.
+
+
+def test_get_auth_callbacks():
+    def A(_: OIDCUserModel) -> bool:
+        return True
+
+    def B(_: OIDCUserModel) -> bool:
+        return True
+
+    def C(_: OIDCUserModel) -> bool:
+        return True
+
+    def D(_: OIDCUserModel) -> bool:
+        return True
+
+    @step("bar")
+    def bar():
+        return {}
+
+    @step("baz")
+    def baz():
+        return {}
+
+    workflow = make_workflow(
+        f=lambda: {},
+        description="description",
+        initial_input_form=None,
+        target=Target.SYSTEM,
+        steps=StepList([]),
+        authorize_callback=None,
+        retry_auth_callback=None,
+    )
+
+    cases = [
+        ((None, None, None, None), (None, None)),
+        ((A, None, None, None), (A, A)),
+        ((None, B, None, None), (None, B)),
+        ((A, B, None, None), (A, B)),
+        ((None, None, C, None), (C, C)),
+        ((A, None, C, None), (C, C)),
+        ((None, B, C, None), (C, C)),
+        ((A, B, C, None), (C, C)),
+        ((None, None, None, D), (None, D)),
+        ((A, None, None, D), (A, D)),
+        ((None, B, None, D), (None, D)),
+        ((A, B, None, D), (A, D)),
+        ((None, None, C, D), (C, D)),  # 4
+        ((A, None, C, D), (C, D)),
+        ((None, B, C, D), (C, D)),
+        ((A, B, C, D), (C, D)),
+    ]
+    for case in cases:
+        auth, retry, step_resume_auth, step_retry_auth = case[0]
```

Comment:
You could use `@pytest.mark.parametrize()` to run each case as a different input scenario for this unittest. Benefit is that you don't need a for loop, you can see whether it's just 1 scenario failing or 3 (or all). There's several other tests in the codebase that utilize this that you can use as a reference.

If there's test preparations that are relatively "expensive", they could be moved into a module-level fixture (`@pytest.fixture(scope="module")`)

Reply:
Thanks Mark! I added this in, and it immediately improved my testing experience.